lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,909 @@
+    rewriter.create<ConnectOp>(oldOp.getLoc(),
+                               topModuleOp.getArgument(numArg - e + i),
+                               oldOp.getOperand(i));
+  rewriter.eraseOp(&oldOp);
+}
+
+//===----------------------------------------------------------------------===//
+// HandshakeToFIRRTL lowering Pass
+//===----------------------------------------------------------------------===//
+
+/// Process of lowering:
+///
+/// 0)  Create and go into a new FIRRTL circuit;
+/// 1)  Create and go into a new FIRRTL top-module;
+/// 2)  Inline Handshake FuncOp region into the FIRRTL top-module;
+/// 3)  Traverse and convert each Standard or Handshake operation:
+///   i)    Check if an identical sub-module exists. If so, skip to vi);
+///   ii)   Create and go into a new FIRRTL sub-module;
+///   iii)  Extract data (if applied), valid, and ready subfield from each port
+///         of the sub-module;
+///   iv)   Build combinational logic;
+///   v)    Exit the sub-module and go back to the top-module;
+///   vi)   Create an new instance for the sub-module;
+///   vii)  Connect the instance with its predecessors and successors;
+/// 4)  Erase the Handshake FuncOp.
+///
+/// createTopModuleOp():  1) and 2)
+/// checkSubModuleOp():   3.i)
+/// createSubModuleOp():  3.ii)
+/// extractSubfields():   3.iii)
+/// build*Logic():        3.iv)
+/// createInstOp():       3.v), 3.vi), and 3.vii)
+///
+/// Please refer to test_addi.mlir test case.
+struct HandshakeFuncOpLowering : public OpConversionPattern<handshake::FuncOp> {
+  using OpConversionPattern<handshake::FuncOp>::OpConversionPattern;
+
+  LogicalResult match(Operation *op) const override { return success(); }
+
+  void rewrite(handshake::FuncOp funcOp, ArrayRef<Value> operands,
+               ConversionPatternRewriter &rewriter) const override {
+    // Create FIRRTL circuit and top-module operation.
+    CircuitOp circuitOp = rewriter.create<CircuitOp>(
+        funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()));
+    rewriter.setInsertionPointToStart(circuitOp.getBody());
+    FModuleOp topModuleOp = createTopModuleOp(funcOp, rewriter);
+
+    // Traverse and convert each operation in funcOp.
+    for (Operation &op : topModuleOp.getBody().front()) {
+      if (isa<handshake::ReturnOp>(op))
+        convertReturnOp(op, topModuleOp, rewriter);
+
+      // This branch takes care of all non-timing operations that require to
+      // be instantiated in the top-module.
+      else if (op.getDialect()->getNamespace() != StringRef("firrtl")) {
+        FModuleOp subModuleOp = checkSubModuleOp(topModuleOp, op);
+
+        // Check if the sub-module already exists.
+        if (!subModuleOp) {
+          subModuleOp = createSubModuleOp(topModuleOp, op, rewriter);
+
+          Operation *termOp = subModuleOp.getBody().front().getTerminator();
+          Location insertLoc = termOp->getLoc();
+          rewriter.setInsertionPoint(termOp);
+
+          ValueVectorList portList =
```

Comment:
It would be nice if the portList could be a data structure which was typesafe when passed to the build...() functions below.

Reply:
+1, that could be split out as a separate refactoring in a follow-on patch though.